cpp/daal/src/externals/service_rng_ref.h

```diff
@@ -0,0 +1,363 @@
+    virtual int uniformRNG(const size_t n, float * r, const float a, const float b, const int method)         = 0;
+    virtual int uniformRNG(const size_t n, double * r, const double a, const double b, const int method)      = 0;
+    virtual int gaussianRNG(const size_t n, float * r, const float a, const float sigma, const int method)    = 0;
+    virtual int gaussianRNG(const size_t n, double * r, const double a, const double sigma, const int method) = 0;
+    virtual int bernoulliRNG(const size_t n, int * r, const double p, const int method)                       = 0;
+    virtual int getSize()                                                                                     = 0;
+    virtual void clone(void * dest) const                                                                     = 0;
+    virtual StateIface * clone() const                                                                        = 0;
+    virtual void assign(const void * src)                                                                     = 0;
+    virtual void discard(size_t nSkip)                                                                        = 0;
+};
+
+template <typename Gen = std::mt19937>
+class State : public StateIface
+{
+    using ThisType = State<Gen>;
+
+public:
+    int uniformRNG(const size_t n, size_t * r, const size_t a, const size_t b, const int method) override { return iUniform(n, r, a, b, method); }
+
+    int uniformRNG(const size_t n, int * r, const int a, const int b, const int method) override { return iUniform(n, r, a, b, method); }
+
+    int uniformRNG(const size_t n, float * r, const float a, const float b, const int method) override { return fUniform(n, r, a, b, method); }
+
+    int uniformRNG(const size_t n, double * r, const double a, const double b, const int method) override { return fUniform(n, r, a, b, method); }
+
+    int gaussianRNG(const size_t n, float * r, const float a, const float sigma, const int method) override
+    {
+        return fGaussian(n, r, a, sigma, method);
+    }
+
+    int gaussianRNG(const size_t n, double * r, const double a, const double sigma, const int method) override
+    {
+        return fGaussian(n, r, a, sigma, method);
+    }
+
+    int bernoulliRNG(const size_t n, int * r, const double p, const int method) override { return iBernoulli(n, r, p, method); }
+
+    State(const int brngId, const size_t seedSize, const unsigned int * seed)
+        : _gen(0), _brngId(brngId), _seedSize(seedSize), _seed(nullptr), _nSkip(0)
+    {
+        _seed = (unsigned int *)daal::services::daal_malloc(sizeof(unsigned int) * _seedSize);
+        if (_seedSize > 0 && seed != nullptr)
+        {
+            for (size_t i = 0; i < _seedSize; i++)
+            {
+                _seed[i] = seed[i];
+            }
+            _gen.seed(_seed[0]);
+        }
+    }
+    State(const State & other) : _gen(0), _brngId(other._brngId), _seedSize(other._seedSize), _seed(nullptr), _nSkip(other._nSkip)
```

Comment:
Same comment as above.
And _nSkip is not needed in case the standard way of serialization/deserialization is used.

Reply:
_nSkip is necesssary at least to emulate SkipAhead by discard and to track a number of discarded elements

Reply:
There is really no need in tracking the number of discarded elements. Because it is already encoded in the state of std::mt19937. Storing nSkip as a separate field duplicates the info already available in std::mt19937 RNG's state. You can just have nSkip as an input parameter of SkipAhead method, and do gen.discard(nSkip) in the implementation.